chall08/pcuadrad.py

```diff
@@ -0,0 +1,185 @@
+                    pass
+                elif square == 'E':
+                    limit += 1
+                    if not self.exitMark:
+                        self.setExit(pos, posy)
+                    else:
+                        return False
+                elif square == '*':
+                    self.setMine(pos, posy)
+                elif square == 'M':
+                    if not self.robotMark:
+                        self.setRobot(pos, posy)
+                    else:
+                        return False
+                else:
+                    return False
+        return True if limit == 2 else False
+
+    def constructMap(self, mapi):
+        '''Create the map in a two-dimensional array and check for errors.'''
+        self.invalidMap = False
+        for line, linePos in zip(mapi, range(len(mapi) - 1)):
+            if linePos > 0 and len(line) != len(lastLine):
+                self.invalidMap = True
+            if self.verifyLine(line, linePos, len(mapi)):
+                self.map.append([char for char in line])
+            else:
+                self.invalidMap = True
+            lastLine = line
+        if not self.robotMark or not self.exitMark:
+            self.invalidMap = True
+
+    def makeRobotMove(self, befX, befY, newX, newY):
+        '''Check if the robot could move to the next x-y position.\
+            Also checks in the new position if stepped a mine.'''
+        if self.map[newX][newY] != '+' and self.robot.active:
+            self.robot.updatePosition(newX, newY)
+            self.map[befX][befY] = '0'
+            self.map[newX][newY] = 'M'
+            die, xMine, yMine = self.robot.explodes(self.mines)
+            if die:
+                sys.exit("Exploded.")
+
+    def moveRobot(self, action: chr):
+        '''Check if the command movement is valid and take the corresponding action.'''
+        xBefore = self.robot.x
+        yBefore = self.robot.y
+        if action == 'I':
+            self.robot.startEngine()
+        elif action == 'K':
+            self.robot.stopEngine()
+        elif action == 'N':
+            self.makeRobotMove(xBefore, yBefore, self.robot.x - 1, self.robot.y)
+        elif action == 'S':
+            self.makeRobotMove(xBefore, yBefore, self.robot.x + 1, self.robot.y)
+        elif action == 'E':
+            self.makeRobotMove(xBefore, yBefore, self.robot.x, self.robot.y + 1)
+        elif action == 'W':
+            self.makeRobotMove(xBefore, yBefore, self.robot.x, self.robot.y - 1)
+        else:
+            self.invalidMap = True
+    
+    def CommandsRobot(self, instr: str):
+        '''Receive the instruction string and process them one by one.\
+            If the map is already invalid it does nothing.'''
+        instr = instr.rstrip('\n')
+        if not self.invalidMap:
+            for command in instr:
+                self.moveRobot(command)
+
+def main():
+    game = MineField()
+    mapi = []
+    for line in sys.stdin:
```

Comment:
So. A little trick here, where you wouldn't have had to append every line and construct the board yourself, would have been to use `sys.stdin.read()`, which will hang until `EOF`, and then just `.split()` the board and the instructions between `"\n\n"`, and then split every line of the board with `.split("\n")`, which would generate a list with each row for you.

You will see in the posted solution I use this method.